# ProphageMerger

ProphageMerger is a Snakemake workflow designed to predict prophage regions in bacterial genomes using a combination of tools ([VirSorter2](https://github.com/jiarong/VirSorter2) v2.2.4 + [CheckV](https://bitbucket.org/berkeleylab/checkv/src/master/) v1.0.3, [GeNomad](https://github.com/apcamargo/genomad) v1.11.1 and [VIBRANT](https://github.com/AnantharamanLab/VIBRANT) v1.2.1). CheckV is run after VirSorter2, using the output file generated by VirSorter2. Predicted regions that overlap are merged, with the minimum start coordinate and maximum end coordinate selected for each merged region. The resulting maximized prophage regions are intended for use in the viral tagging analysis. The sequences of the maximized prophage regions are evaluated using CheckV. 

## Set up environment

### Clone the repository 

Approximately 34GB of storage is needed for the installation of tools and databases.

```bash
git clone https://github.com/xwu35/ProphageMerger.git
```

### Install Snakemake

ProphageMerger is built for Snakemake version 7. Version 8 and above introduce breaking changes and deprecations and have not been tested. It may not function correctly with newer versions. Please install Snakemake version 7 using the script below.

```bash
# option 1 using conda
conda env create -n snakemake -f ProphageMerger/snakemake_env.yml

# option 2 using mamba if it's installed
mamba env create -n snakemake -f ProphageMerger/snakemake_env.yml
```

If you have installed Snakemake version 7 on your system without using the commands above, make sure that `Biopython` is also installed. If it is not already installed, activate your snakemake environment and run `pip install biopython==1.86`.

### Download snakemake profile

The profile is required to run the workflow on HPC. Skip this step if you already have a SLURM profile in `~/.config/snakemake`.

```bash
# download the profile
git clone https://github.com/xwu35/slurm

# move the profile to the right directory
mv slurm ~/.config/snakemake 
```

### Export PATH

Add the ProphageMerger path to your environment variable so you can run `ProphageMerger.py` without the full path.

```bash
echo 'export PATH="/your/path/to/ProphageMerger:$PATH"' >> ~/.bashrc
```

To make the changes take effect, you can either log out and log back in, or you can source the `.bashrc` file using the following command:

```bash
source ~/.bashrc
```

## Usage

ProphageMerger requires a FASTA file of the chromosome sequences, which may consist of a single contig or multiple contigs. Detailed usage information can be viewed using the -h or --help flags `ProphageMerger.py -h`.

### Test run

A dry-run can be performed to check which rules will be executed and which files will be produced. 

```bash
conda activate snakemake

ProphageMerger.py --test_run --dryrun
```

Do not run this on the login node. Submit it as an sbatch job. See `run_prophagemerger.sh` for an example, or check the HTCF usage guide here (https://github.com/xwu35/baldridge_lab/blob/main/HTCF.md). 

```bash
conda activate snakemake

ProphageMerger.py --test_run # An output directory named "test_output" will appear
```

### Example run

A dry-run can be performed by specifing the `--dryrun` flag

```bash
conda activate snakemake

ProphageMerger.py \
     -g /path/to/your/genome/sequence/fasta \
     -o /name/your/output/dir
```

## Output description

|                     Filename                       |                     Description                    |
|----------------------------------------------------|----------------------------------------------------|
| `results/all_coordinates.txt`                      | Predicted prophage coordinates from all tools      |
| `results/final_coordinates.txt`                    | Maximized prophage coordinates after merging       | 
| `results/final_coordinates_0-based.bed`            | Same as final_coordinates.txt, but this version has been converted to 0-based coordinates. It can be used as input for [VTPhageFinder](https://github.com/xwu35/VTPhageFinder)                                                  |
| `results/prophage_region_sequences.fa`             | Sequences of the final prophage regions            |
| `results/checkv_evaluation/quality_summary.tsv`    | Prophage region quality                            |
 	

## Merge predicted prophage regions

If the prediction tools have already been executed elsewhere without using this pipeline, and you only need to merge the predicted regions, you can use the `combine_prophage_coordinates.py` script. Detailed usage information can be viewed using the -h or --help flags `/your/path/to/ProphageMerger/workflow/scripts/combine_prophage_coordinates.py -h`.

```bash
conda activate snakemake 

/your/path/to/ProphageMerger/workflow/scripts/combine_prophage_coordinates.py \
     --virsorter2 path/to/final-viral-boundary.tsv \
     --checkv path/to/proviruses.fna  \
     --genomad path/to/prefix_provirus.tsv \
     --vibrant path/to/VIBRANT_integrated_prophage_coordinates_prefix.tsv \
     --all_coordinates all_coordinates.txt \
     --final_coordinates final_coordinates.txt \
     --bed_file final_coordinates_0-based.bed
```
